test/integration/sys-hmac.int.c

```diff
@@ -0,0 +1,141 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/*******************************************************************************
+ * Copyright 2017-2018, Fraunhofer SIT sponsored by Infineon Technologies AG
+ * All rights reserved.
+ *******************************************************************************/
+
+#ifdef HAVE_CONFIG_H
+#include <config.h>
+#endif
+
+#include "tss2_tpm2_types.h"
+#include "tss2_sys.h"
+#include "util/tpm2b.h"
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <inttypes.h>
+
+#include "tss2_tcti.h"
+#include "tss2_sys.h"
+
+#define LOGMODULE test
+#include "util/log.h"
+#include "test.h"
+#include "sys-util.h"
+
+int
+test_sys_hmac(TSS2_SYS_CONTEXT * sys_context)
+{
+
+    TSS2_RC rc;
+    TPM2B_PUBLIC out_public = { 0 };
+    TPM2B_CREATION_DATA creation_data = {0,};
+    creation_data.size = 0;
+    TPM2B_DIGEST creation_hash = {
+        .size = sizeof(creation_hash.buffer)
+    };
+    TPMT_TK_CREATION creation_ticket = { 0 };
+    TPM2B_NAME name;
+    TSS2L_SYS_AUTH_COMMAND  sessions_cmd = {
+        .auths = {{ .sessionHandle = TPM2_RH_PW }},
+        .count = 1
+    };
+    TSS2L_SYS_AUTH_RESPONSE  sessions_rsp     = {
+        .auths = { 0 },
+        .count = 0
+    };
+    TPM2_HANDLE primaryHandle;
+
+    if (sys_context == NULL) {
+        return TSS2_RC_LAYER_MASK | TSS2_BASE_RC_BAD_REFERENCE;
+    }
+
+    TPM2B_DIGEST *outHMAC = NULL;
```

Comment:
TPM2B_DIGEST outHMAC = { 0 };

Reply:
I tried this but I keep on getting a weird segmentation fault

> Execute the test script
../script/int-log-compiler.sh: line 105: 72604 Segmentation fault      (core dumped) env TPM20TEST_TCTI_NAME="${TPM20TEST_TCTI_NAME}" TPM20TEST_SOCKET_ADDRESS="${TPM20TEST_SOCKET_ADDRESS}" TPM20TEST_SOCKET_PORT="${TPM20TEST_SOCKET_PORT}" TPM20TEST_TCTI="${TPM20TEST_TCTI}" TPM20TEST_DEVICE_FILE="${TPM20TEST_DEVICE_FILE}" G_MESSAGES_DEBUG=all ${@: -1}
Script returned 1[39](https://github.com/tpm2-software/tpm2-tss/actions/runs/3332355254/jobs/5513138018#step:4:40)
TPM changed state during test
State before (/workspace/tpm2-tss/build/test/integration/sys-mac.int_state1):
cap0: 000000050000000[40](https://github.com/tpm2-software/tpm2-tss/actions/runs/3332355254/jobs/5513138018#step:4:41)00403ffffff000b03ffffff000c03ffffff000d03ffffff00
cap1: 000000010000001800000000000000010000000200000003000000040000000[50](https://github.com/tpm2-software/tpm2-tss/actions/runs/3332355254/jobs/5513138018#step:4:51)00000060000000700000008000000090000000a0000000b0000000c0000000d0000000e0000000f000000100000001100000012000000130000001400000015000000160000001700
cap2: 000000010000000000
cap3: 000000010000000000
cap4: 000000010000000000
cap5: 000000010000000000
cap6: 000000010000000000
State after (/workspace/tpm2-tss/build/test/integration/sys-mac.int_state2):
cap0: 0000000500000004000403ffffff000b03ffffff000c03ffffff000d03ffffff00
cap1: 0000000100000018000000000000000100000002000000030000000400000005000000060000000700000008000000090000000a0000000b0000000c0000000d0000000e0000000f000000100000001100000012000000130000001400000015000000160000001700
cap2: 000000010000000000
cap3: 000000010000000000
cap4: 00000001000000018000000000
cap5: 000000010000000000
cap6: 000000010000000000
FAIL test/integration/sys-mac.int (exit status: 1)

Someplace in the SAPI complete function it unmarshals the cmd buffer into the outHMAC buffer, maybe thats causing a problem?

Reply:
@malikabhi05 this works for me:\
```diff
diff --git a/test/integration/sys-hmac.int.c b/test/integration/sys-hmac.int.c
index 9ae605b3999f..7ff51d8386ff 100644
--- a/test/integration/sys-hmac.int.c
+++ b/test/integration/sys-hmac.int.c
@@ -52,8 +52,7 @@ test_sys_hmac(TSS2_SYS_CONTEXT * sys_context)
         return TSS2_RC_LAYER_MASK | TSS2_BASE_RC_BAD_REFERENCE;
     }
 
-    TPM2B_DIGEST *outHMAC = NULL;
-    outHMAC = calloc(sizeof(TPM2B_DIGEST), 1);
+    TPM2B_DIGEST outHMAC = { 0 };
 
     TPM2B_SENSITIVE_CREATE in_sensitive = { 0 };
     TPM2B_PUBLIC inPublic = { 0 };
@@ -89,7 +88,7 @@ test_sys_hmac(TSS2_SYS_CONTEXT * sys_context)
                                               1, 2, 3, 4, 5, 6, 7, 8, 9}};
 
     rc = Tss2_Sys_HMAC(sys_context, primaryHandle, &sessions_cmd, &test_buffer,
-                       TPM2_ALG_SHA256, outHMAC, &sessions_rsp);
+                       TPM2_ALG_SHA256, &outHMAC, &sessions_rsp);
 
     if (rc != TSS2_RC_SUCCESS) {
         LOG_ERROR("Tss2_Sys_MAC FAILED! Response Code : 0x%x", rc);
@@ -98,7 +97,7 @@ test_sys_hmac(TSS2_SYS_CONTEXT * sys_context)
     TPMT_SIGNATURE sig;
     sig.signature.hmac.hashAlg = TPM2_ALG_SHA256;
     sig.sigAlg = TPM2_ALG_HMAC;
-    memcpy(sig.signature.hmac.digest.sha256, outHMAC->buffer, outHMAC->size);
+    memcpy(sig.signature.hmac.digest.sha256, outHMAC.buffer, outHMAC.size);
 
     TPM2B_DIGEST dig = { .size = 20,
                          .buffer={0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 0,
```